File: Advanced/PageAndRDL/CustomDataProvider/VB.NET/CustomDataProviderUI/QueryEditor.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing.Design
Imports System.IO
Imports System.Text.RegularExpressions
Imports System.Windows.Forms.Design

Public NotInheritable Class QueryEditor
	Inherits UITypeEditor

	Public Overloads Overrides Function GetEditStyle(ByVal context As System.ComponentModel.ITypeDescriptorContext) As UITypeEditorEditStyle
		Return UITypeEditorEditStyle.DropDown
	End Function

	Public Overloads Overrides Function EditValue(ByVal context As System.ComponentModel.ITypeDescriptorContext, ByVal provider As System.IServiceProvider, ByVal value As Object) As Object
		Dim edSvc As IWindowsFormsEditorService = CType(provider.GetService(GetType(IWindowsFormsEditorService)), IWindowsFormsEditorService)
		Dim fileSelector As New CSVFileSelector()
		edSvc.DropDownControl(fileSelector)
		Dim csvFileName As String = fileSelector.CSVFileName
		If String.IsNullOrEmpty(csvFileName) Then
			Return String.Empty
		End If
		If Not File.Exists(csvFileName) Then
			Return String.Empty
		End If
		Return GetCSVQuery(csvFileName)
	End Function

	Private Shared Function GetCSVQuery(ByVal csvFileName As String) As String
		Dim sr As StreamReader = Nothing
		Try
			sr = New StreamReader(csvFileName)
			Dim ret As String = String.Empty
			Dim currentLine As String
			Dim line As Integer = 0

			currentLine = sr.ReadLine()
			While (currentLine) IsNot Nothing
				If (line).Equals(0) Then
					ret += ProcessColumnsDefinition(currentLine) + vbCr & vbLf
				Else
					ret += currentLine + vbCr & vbLf
				End If
				line = line + 1
				currentLine = sr.ReadLine()
			End While
			Return ret
		Catch generatedExceptionName As IOException
			Return String.Empty
		Finally
			If Not (sr) Is Nothing Then
				sr.Close()
			End If
		End Try
	End Function

	Private Shared Function ProcessColumnsDefinition(ByVal line As String) As String
		Const ColumnWithDataTypeRegex As String = "[""]?\w+[\""]?\(.+\)"
		Dim columns As String() = line.Split(New String() {","}, StringSplitOptions.None)
		Dim ret As String = Nothing
		For Each column As String In columns
			If Not String.IsNullOrEmpty(ret) Then
				ret += ","
			End If
			If Not Regex.Match(column, ColumnWithDataTypeRegex).Success Then
				ret += column + "(string)"
			Else
				ret += column
			End If
		Next
		Return ret
	End Function

End Class

File: Advanced/PageAndRDL/CustomDataProvider/C#/CustomDataProviderUI/QueryEditor.cs
using System;
using System.Drawing.Design;
using System.IO;
using System.Text.RegularExpressions;
using System.Windows.Forms.Design;

namespace GrapeCity.ActiveReports.Samples.CustomDataProviderUI
{
	public sealed class QueryEditor : UITypeEditor
	{
		public override UITypeEditorEditStyle GetEditStyle(System.ComponentModel.ITypeDescriptorContext context)
		{
			return UITypeEditorEditStyle.DropDown;
		}
		
		public override object EditValue(System.ComponentModel.ITypeDescriptorContext context, System.IServiceProvider provider, object value)
		{
			IWindowsFormsEditorService edSvc = (IWindowsFormsEditorService)provider.GetService(typeof(IWindowsFormsEditorService));
			CSVFileSelector fileSelector = new CSVFileSelector();
			edSvc.DropDownControl(fileSelector);
			string csvFileName =  fileSelector.CSVFileName;
			if(string.IsNullOrEmpty(csvFileName)) return string.Empty;
			if(!File.Exists(csvFileName)) return string.Empty;
			return GetCSVQuery(csvFileName);
		}
		
		/// <summary>
		/// Reads the content of the specified file and builds the CSV data provider query string.
		/// </summary>
		/// <param name="csvFileName">The full path of the CSV file.</param>
		/// <returns>CSV data provider query string.</returns>
		/// <remarks>
		/// This method just reads the file lines one by one and adds them to the query string.
		/// The first line of CSV query string should have the columns names and columns data types,
		/// so we run the special process for the first line.
		/// </remarks>
		private static string GetCSVQuery(string csvFileName)
		{
			StreamReader sr = null;
			try
			{
				sr = new StreamReader(csvFileName);
				string ret = string.Empty;
				string currentLine;
				int line = 0;
				while ((currentLine = sr.ReadLine()) != null)
				{
					if (line == 0)
						ret += ProcessColumnsDefinition(currentLine) + "\r\n";
					else
						ret += currentLine + "\r\n";
					line++;
				}
				return ret;
			}
			catch (IOException)
			{
				return string.Empty;
			}
			finally
			{
				if (sr != null)
					sr.Close();
			}
		}

		/// <summary>
		///Reads the CSV data columns definition from the specified string and adjusts it if necessary.
		/// </summary>
		/// <param name="line">The string that contains the CSV data columns definition.</param>
		/// <returns>The CSV data columns definition that includes the data types definition.</returns>
		private static string ProcessColumnsDefinition(string line)
		{
			const string ColumnWithDataTypeRegex = @"[""]?\w+[\""]?\(.+\)";
			string[] columns = line.Split(new string[] {","}, StringSplitOptions.None);
			string ret = null;
			foreach(string column in columns)
			{
				if(!string.IsNullOrEmpty(ret))
					ret+= ",";
				if(!Regex.Match(column, ColumnWithDataTypeRegex).Success)
				{
					ret += column + "(string)";
				}
				else
				{
					ret += column;
				}
			}
			return ret;
		}
	}
}
